Dapper.MoqTests/ParametersObjectBuilder.cs

```diff
@@ -45,6 +47,94 @@ public object FromParameters(IReadOnlyCollection<DbParameter> parameters)
             return instance;
         }
 
+        private IEnumerable<DbParameter> ReGroupSerialisedParameters(IReadOnlyCollection<DbParameter> parameters)
+        {
+            var serialisedParameters = parameters.Where(IsSerialisedParameter).ToArray();
+            return parameters.Except(serialisedParameters).Concat(ReconstituteParameters(serialisedParameters));
+        }
+
+        private IEnumerable<DbParameter> ReconstituteParameters(IEnumerable<DbParameter> serialisedParameters)
+        {
+            var orderedParameters = serialisedParameters.OrderBy(p => p.ParameterName);
+            var buffer = new List<DbParameter>();
+            string parameterName = null;
+
+            foreach (var parameter in orderedParameters)
+            {
+                if (!buffer.Any())
+                {
+                    buffer.Add(parameter);
+                    parameterName = GetParameterName(parameter.ParameterName);
+                    continue;
+                }
+
+                if (parameter.ParameterName == parameterName + (buffer.Count + 1))
+                {
+                    buffer.Add(parameter);
+                    continue;
+                }
+
+                yield return new MockDbParameter
+                {
+                    ParameterName = parameterName,
+                    Value = GetReconstitutedParameterValue(buffer.Select(p => p.Value).ToArray())
+                };
+                parameterName = null;
+                buffer.Clear();
+                buffer.Add(parameter);
+                parameterName = GetParameterName(parameter.ParameterName);
+            }
+
+            if (buffer.Any())
+            {
+                yield return new MockDbParameter
+                {
+                    ParameterName = parameterName,
+                    Value = GetReconstitutedParameterValue(buffer.Select(p => p.Value).ToArray())
+                };
+            }
+        }
+
+        private static object GetReconstitutedParameterValue(IReadOnlyCollection<object> values)
+        {
+            var firstValueType = values.First(v => v != null)?.GetType();
+            if (firstValueType == null)
+                return values; //all items are null
+
+            if (values.All(v => v != null && v.GetType().Equals(firstValueType)))
+            {
+                //assumes that the first item is of the 'widest' type
+                //the the value should be an array of <firstValueType> rather than an array of <object>
+                var typedArray = Array.CreateInstance(firstValueType, values.Count);
+                for (var itemIndex = 0; itemIndex < values.Count; itemIndex++)
+                {
+                    typedArray.SetValue(values.ElementAt(itemIndex), itemIndex);
+                }
+
+                return typedArray;
+            }
+
+            if (values.Any(v => v == null))
```

Comment:
Just because some of the values are null doesn't mean the array will be typed to object. It could, for example be an array of various versions of a complex type - types that all inherent/implement an common type.

Reply:
On reflection we probably don't need worry about any collection type except `IEnumerable<object>` as the parameters comparer will always cast to this type before comparing the collection.

These changes should be simplified into `Cast<object>()`